USB2 INNO PHY 驱动修改

185264646 opened the pull request:
海思原来写的代码把PHY和测试总线驱动混一块了，比较好的做法应该是分开成两部分。`hisi_inno_phy_write_reg`函数拆分到单独的总线驱动里面去，phy驱动只保留`hisi_inno_phy_setup()`逻辑就可以了。我打算把驱动修改成这种结构，接下来添加其他设备支持会容易一些。

@yangfl 如果有空的话，能否帮忙测试一下这一版本的驱动能否在mv100上正常运行？设备树修改可以直接看提交记录里的binding。

当然，测试完后会加上你的Tested-by标签。

能测试一下读取函数能否正常工作的话更好，可惜这需要再写代码了。

-----

重新revert回原来的版本了，方便的话可以review一下。

185264646 committed:
phy: hisi-inno-phy: add support for hi3798mv200-usb2-phy

This should be considered a dirty hack. The proper solution would be
extracting write_reg logic to a separate regmap driver. Leaving only
"write BIT(2) to address 0x6" to the PHY driver.

The initial commit is already doing things wrong. The following patches
adding hi3798mv100 support is also very confusing. The name of the
enumeration "PHY_TYPE_x" is very misleading as if it's the phy which is
different across SoCs. But actually it's the bus (i.e. how to write to a
given address) which is different, not the PHY.

# Describe the purpose of this series. The information you put here
# will be used by the project maintainer to make a decision whether
# your patches should be reviewed, and in what priority order. Please be
# very detailed and link to any relevant discussions or sites that the
# maintainer can review to better understand your proposed changes. If you
# only have a single patch in your series, the contents of the cover
# letter will be appended to the "under-the-cut" portion of the patch.

# Lines starting with # will be removed from the cover letter. You can
# use them to add notes or reminders to yourself. If you want to use
# markdown headers in your cover letter, start the line with ">#".

# You can add trailers to the cover letter. Any email addresses found in
# these trailers will be added to the addresses specified/generated
# during the b4 send stage. You can also run "b4 prep --auto-to-cc" to
# auto-populate the To: and Cc: trailers based on the code being
# modified.

To: Vinod Koul <vkoul@kernel.org>
To: Kishon Vijay Abraham I <kishon@kernel.org>
To: Rob Herring <robh+dt@kernel.org>
To: Krzysztof Kozlowski <krzysztof.kozlowski+dt@linaro.org>
To: Conor Dooley <conor+dt@kernel.org>
To: Jiancheng Xue <xuejiancheng@hisilicon.com>
To: Shawn Guo <shawn.guo@linaro.org>
To: Philipp Zabel <p.zabel@pengutronix.de>
Cc:  <linux-phy@lists.infradead.org>
Cc:  <devicetree@vger.kernel.org>
Cc:  <linux-kernel@vger.kernel.org>
Cc: Kishon Vijay Abraham I <kishon@ti.com>
Cc: David Yang <mmyangfl@gmail.com>
Signed-off-by: Yang Xiwen <forbidden405@outlook.com>
---
Changes in v4:
- split phy test bus operation to a dedicated driver, now it's clear PHY is compatible.
- split YAML convertion into two commits, the other add mv100 compatible (Krzysztof Kozlowski)
- Link to v3: https://lore.kernel.org/r/20240220-inno-phy-v3-0-893cdf8633b4@outlook.com

Changes in v3:
- address a few binding issue mistakenly missing in v2 (Krzysztof Kozlowski)
  - add msg about hi3798mv100 being added to compatible list
  - remove minItems for compatible
  - remove | for reg:
- fix existing dts (hi3798cv200.dtsi) due to binding change.
- Link to v2: https://lore.kernel.org/r/20240217-inno-phy-v2-0-3bf7e87b0e9e@outlook.com

Changes in v2:
- rewrite commit msg to show why hisilicon,hi3798mv100-usb2-phy is added during YAML convertion.
- split required: to multiple line
- add allOf to wrap if:
- remove perictrl wrapper and the second phy in the example
- tested the binding both for mv200 and cv200 dts. fix some silly errors.
- remove Pengcheng Li from To:
Above all are suggested by Krzysztof
- use reset_control_array_* APIs to ensure all resets are controlled
- Link to v1: https://lore.kernel.org/r/20240216-inno-phy-v1-0-1ab912f0533f@outlook.com

--- b4-submit-tracking ---
# This section is used internally by b4 prep for tracking purposes.
{
  "series": {
    "revision": 4,
    "change-id": "20240216-inno-phy-a2d872f6b74b",
    "prefixes": [
      "RFC"
    ],
    "history": {
      "v1": [
        "20240216-inno-phy-v1-0-1ab912f0533f@outlook.com"
      ],
      "v2": [
        "20240217-inno-phy-v2-0-3bf7e87b0e9e@outlook.com"
      ],
      "v3": [
        "20240220-inno-phy-v3-0-893cdf8633b4@outlook.com"
      ]
    }
  }
}

185264646 committed:
dt-bindings: phy: hisi-inno-usb2: convert to YAML

convert the legacy text binding to YAML. While at it, remove
"hisilicon,inno-usb2-phy" fallback compatible.

Signed-off-by: Yang Xiwen <forbidden405@outlook.com>

185264646 committed:
phy: hisilicon: hisi-inno-phy: enable clocks for every ports

This is needed for port1 to work.

Fixes: ba8b0ee81fbb ("phy: add inno-usb2-phy driver for hi3798cv200 SoC")
Signed-off-by: Yang Xiwen <forbidden405@outlook.com>

185264646 committed:
phy: hisilicon: hisi-inno-phy: add support for Hi3798MV200 INNO PHY

Direct MMIO resgiter access is used by Hi3798MV200. For other models,
of_iomap() returns NULL due to insufficient length. So they are
unaffected.

Also Hi3798MV200 INNO PHY has an extra reset required to be deasserted,
switch to reset_control_bulk_() APIs for that.

Signed-off-by: Yang Xiwen <forbidden405@outlook.com>

yangfl commented:
ok，我目前在外，下个月能做

yangfl commented:
这未免有点太小题大做了。目前test bus完全看不出在其他类似设备上有应用，或者有进一步phy内部寄存器的资料，真要有更多用途再抽象成regmap感觉更合适。

185264646 commented:
> 这未免有点太小题大做了。目前test bus完全看不出在其他类似设备上有应用，或者有进一步phy内部寄存器的资料，真要有更多用途再抽象成regmap感觉更合适。

我刚开始就是把这一坨全放到PHY驱动里，结果跟这位maintainer扯了半天：https://lore.kernel.org/all/20240220-inno-phy-v3-0-893cdf8633b4@outlook.com/

实在说服不了他，只好把bus驱动单独写出来了。

yangfl commented:
我感觉你是在yaml误导了他也被他误导了，你的yaml给人感觉就像是有个bus一样，但这个bus其实没必要在dts里体现。
最好还是先看看其他部分，等他忘了这事你再提交一个简单的txt实现方案。

185264646 commented:
> 我感觉你是在yaml误导了他也被他误导了，你的yaml给人感觉就像是有个bus一样，但这个bus其实没必要在dts里体现。 最好还是先看看其他部分，等他忘了这事你再提交一个简单的txt实现方案。

我第一版驱动就这么弄的，结果binding实在糊弄不过去。

这么搞mv200支持不好弄。mv200的inno phy是直接mmio就可以的，设备树怎么写呢？下面的phy port节点的reg到底是写成所谓的phy index还是写成物理地址空间呢？

mv200还多一个reset，这个reset在cv200和mv100上直接由驱动内部代劳了，就是所谓的TEST_PORT_RSTN。但是在mv200上是挂在crg下面的，所以肯定要写进设备树里，然后在驱动里获取。

当然将错就错也可以，就直接假装没有这回事，再给mv200加一个PHY_TYPE_2，然后binding直接别管，邮件不要给这位maintainer发了。试试看能不能糊弄过去。只要phy那边把patch收了，就能进主线。但是我还是希望能把这一块弄清晰一点。

要是有空的话还是帮忙测试一下吧，先发出去试试水，写都写了，说不要就不要也不好。

yangfl commented:
行，过几天有设备看看

yangfl commented:
638ca8b96876f302fe4eb4f41ef570ac8b5e9f51

[   15.847568] platform f9890000.usb: deferred probe pending: platform: wait for supplier /soc@f0000000/peripheral-controller@8a20000/usb2-phy-bus@124/usb2phy/phy@1
[   15.862021] platform f9930000.usb: deferred probe pending: platform: wait for supplier /soc@f0000000/peripheral-controller@8a20000/usb2-phy-bus@158/usb2phy/phy@0
[   15.876474] platform f9880000.usb: deferred probe pending: platform: wait for supplier /soc@f0000000/peripheral-controller@8a20000/usb2-phy-bus@124/usb2phy/phy@1
[   15.890921] platform f9920000.usb: deferred probe pending: platform: wait for supplier /soc@f0000000/peripheral-controller@8a20000/usb2-phy-bus@158/usb2phy/phy@0

185264646 commented:
> [638ca8b](https://github.com/histb-mainline/linux/commit/638ca8b96876f302fe4eb4f41ef570ac8b5e9f51)
> 
> [ 15.847568] platform f9890000.usb: deferred probe pending: platform: wait for supplier /soc@f0000000/peripheral-controller@8a20000/usb2-phy-bus@124/usb2phy/phy@1 [ 15.862021] platform f9930000.usb: deferred probe pending: platform: wait for supplier /soc@f0000000/peripheral-controller@8a20000/usb2-phy-bus@158/usb2phy/phy@0 [ 15.876474] platform f9880000.usb: deferred probe pending: platform: wait for supplier /soc@f0000000/peripheral-controller@8a20000/usb2-phy-bus@124/usb2phy/phy@1 [ 15.890921] platform f9920000.usb: deferred probe pending: platform: wait for supplier /soc@f0000000/peripheral-controller@8a20000/usb2-phy-bus@158/usb2phy/phy@0

能找一找原因吗?尤其是帮忙看看bus驱动probe是否成功了。

yangfl commented:
大概是clocks没开

185264646 commented:
> 大概是clocks没开

哪一个clock呢？可以把相关的dt node发出来研究一下

yangfl commented:
上面那个commit

185264646 committed:
dt-bindings: phy: hisilicon,inno-usb2-phy: add support for Hi3798MV100 INNO PHY

Hi3798MV100 also has a similar INNO USB2 PHY with slightly different
register fields offsets. Document it in the binding.

Signed-off-by: Yang Xiwen <forbidden405@outlook.com>

185264646 committed:
dt-bindings: phy: hisi-inno-usb2: add compatible of hisilicon,hi3798mv200-usb2-phy

Hi3798MV200 INNO USB2 PHY is attached directly to system bus. Add
compatible "hisilicon,hi3798mv200-usb2-phy" for it.

Signed-off-by: Yang Xiwen <forbidden405@outlook.com>

185264646 commented:
@yangfl 现在这个版本如何呢？驱动本身的修改应该对于mv100和cv200是无感知的，而且应该更容易说服binding的maintainer。

yangfl commented:
这个应该是没有任何问题
